// react-ui-dropdown/lib/components/SelectedItem.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SelectedItem = (function (_Component) {
  _inherits(SelectedItem, _Component);

  function SelectedItem() {
    _classCallCheck(this, SelectedItem);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(SelectedItem).apply(this, arguments));
  }

  _createClass(SelectedItem, [{
    key: "getId",
    value: function getId() {
      return this.props.idPrefix + "selected-item";
    }
  }, {
    key: "render",
    value: function render() {
      return _react2.default.createElement(
        "div",
        { className: "dropdown-selected-item", id: this.getId() },
        _react2.default.createElement(
          "span",
          { className: "dropdown-item-title" },
          this.props.title
        ),
        _react2.default.createElement(
          "button",
          {
            className: "dropdown-item-close",
            type: "button",
            onClick: this.props.onClick },
          String.fromCharCode(215)
        )
      );
    }
  }]);

  return SelectedItem;
})(_react.Component);

exports.default = SelectedItem;

SelectedItem.propTypes = {
  idPrefix: _react.PropTypes.string.isRequired,
  title: _react.PropTypes.string.isRequired,
  onClick: _react.PropTypes.func.isRequired
};
